### DISCUSSION TITLE
API Documentation Priority List

### DISCUSSION THREAD

**Discussion by michaelweinold** [brightway-lca/brightway-documentation#66] [Strategy and Planning]
# `brightway2-io`

## Priorities

| symbol | description |
| ------- | ----------- |
| 1️⃣ | high priority |
| 2️⃣ | medium priority |
| 3️⃣ | low priority |
| ❌ | deprecated |
| ⏳ | wait for strategic code changes | 

### core modules (`backup`, etc.) 
- [x] 2️⃣ backup.py : https://github.com/brightway-lca/brightway2-io/pull/169/
- [ ] ❌ chemidplus.py 
- [ ] 2️⃣ compatibility.py 
- [ ] 2️⃣ download_utils.py: https://github.com/brightway-lca/brightway2-io/pull/169/
- [x] 2️⃣ errors.py
- [x] 2️⃣ utils.py: https://github.com/brightway-lca/brightway2-io/pull/169/
- [ ] 3️⃣ modified_database.py
- [ ] 3️⃣ package.py
- [ ] 3️⃣ errors.py
- [ ] 3️⃣ units.py
- [ ] 3️⃣ unlinked_data.py
- [ ] 3️⃣ utils.py

### `export`
- [x] 2️⃣ csv.py: https://github.com/brightway-lca/brightway2-io/pull/169/
- [x] 2️⃣ excel.py: https://github.com/brightway-lca/brightway2-io/pull/169/
- [x] 2️⃣ gexf.py: https://github.com/brightway-lca/brightway2-io/pull/169/
- [x] 2️⃣ matlab.py: https://github.com/brightway-lca/brightway2-io/pull/169/

### `extractors`
- [ ] 1️⃣ csv.py 
- [ ] ❌ ecospold1.py: https://github.com/brightway-lca/brightway2-io/pull/176
- [ ] ❌ ecospold1_lcia.py
- [ ] ❌ ecospold2.py
- [x] 1️⃣ excel.py: https://github.com/brightway-lca/brightway2-io/pull/182
- [ ] ⏳ exiobase.py
- [ ] 1️⃣ json_ld.py
- [ ] 1️⃣ simapro_csv.py
- [ ] 1️⃣ simapro_lcia_csv.py`

### importers
- [x] 1️⃣ base.py: https://github.com/brightway-lca/brightway2-io/pull/172
- [ ] 1️⃣ base_lci.py 
- [ ] 1️⃣ base_lcia.py
- [x] ⏳ ecoinvent_lcia.py: https://github.com/brightway-lca/brightway2-io/pull/166
- [ ] ⏳ ecospold1.py
- [ ] ⏳ ecospold1_lcia.py
- [x] ⏳ ecospold2.py: https://github.com/brightway-lca/brightway2-io/pull/172
- [x] ⏳ ecospold2_biosphere.py: https://github.com/brightway-lca/brightway2-io/pull/166
- [ ] 1️⃣ excel.py 
- [x] 2️⃣ excel_lcia.py: https://github.com/brightway-lca/brightway2-io/pull/166
- [ ] ❌ exiobase3_hybrid.py
- [ ] ❌ exiobase3_monetary.py
- [ ] 2️⃣ json_ld.py
- [x] 2️⃣ json_ld_lcia.py: https://github.com/brightway-lca/brightway2-io/pull/166
- [ ] 2️⃣ simapro_csv.py
- [x] 2️⃣ simapro_lcia_csv.py: https://github.com/brightway-lca/brightway2-io/pull/166

### strategies 
- [ ] 1️⃣ biosphere.py 
- [ ] 1️⃣ csv.py
- [ ] ❌ ecospold1.py
- [ ] ❌ ecospold1_allocation.py
- [x] ❌ ecospold2.py: https://github.com/brightway-lca/brightway2-io/pull/175
- [x] ❌ exiobase.py: https://github.com/brightway-lca/brightway2-io/pull/177 
- [x] 1️⃣ generic.py: https://github.com/brightway-lca/brightway2-io/pull/178
- [x] 1️⃣ json_ld.py: https://github.com/brightway-lca/brightway2-io/pull/180
- [ ] 1️⃣ json_ld_allocation.py
- [ ] 1️⃣ json_ld_lcia.py
- [ ] 1️⃣ lcia.py
- [ ] 1️⃣ locations.py
- [ ] 1️⃣ migrations.py
- [ ] 1️⃣ simapro.py
- [ ] 1️⃣ special.py
- [ ] 1️⃣ useeio.py

**Comment by michaelweinold** [comment]
# Freelancer Performance Estimate

## Performance Calculation

### @Cerebrovinny (as of 19.04.2023)

| PR | # classes | # functions | # lines of doc. | comment | 
| --- | --------- | ----------- | --------------- | ---------- |
| https://github.com/brightway-lca/brightway2-io/pull/177 | 0 | 8 | ~200 | includes `Examples` section |
| https://github.com/brightway-lca/brightway2-io/pull/175 | 0 | 18 | ~350 | includes `Examples` section |
| https://github.com/brightway-lca/brightway2-io/pull/171 | 0 | 15 | ~150 | includes `Examples` section | 
| https://github.com/brightway-lca/brightway2-io/pull/184 | 0 | 1 | minor | |
| https://github.com/brightway-lca/brightway2-io/pull/182 | 1 | 4 | ~100 | includes `Examples` section | 
| https://github.com/brightway-lca/brightway2-io/pull/181 | 1 | 0 | ~50 | includes `Examples` section |
| https://github.com/brightway-lca/brightway2-io/pull/178 | 0 | 10 | ~600 | includes `Examples` section |
| https://github.com/brightway-lca/brightway2-io/pull/180 | 0 | 10 | ~600 | includes `Examples` section |

| Corresponding time logged | `57:50h` |
| --------------------------- | --------- |

__ca. 15.-2.5 functions/hr__

## Estimate (Time Remaining)

### [`brightway2-io`](https://github.com/brightway-lca/brightway2-io)

| description | number (estimate) |
| ----------- | -------- |
| functions | 640 |
| functions with trivial documentation | 50 |
| functions already documented well | 50 |
| functions documented by @Cerebrovinny | 40 |
| __remaining functions__ | __500__ |

Estimate: (50 functions / 19:00h) ~ `2.6 functions/h`
Estimate: `192h` for remaining functions

**Reply by cmutel** [reply]
Wow, this is amazing work, thanks a lot @michaelweinold.

A few inputs for prioritization:

* We are going to switch to [pyecospold](https://github.com/sami-m-g/pyecospold) for ecospold 1/2 **extraction**, so these extractors don't need documentation. I would also strongly deprioritize the **importers** as they will need to change a lot for the new extractor.
* Please also deprioritize exiobase stuff, this is a collection of hacks that needs to either be replaced completely or at least be rewritten.
* I would boost SimaPro CSV and JSON LD, as these are places where people will read the documentation
* `chemidplus.py` can be ignored

I would therefore like to commit to the following, in order of priority, and then we can decide if we want to continue with this project and `bw2io`:

* `importers/base_lci.py`
* `importers/base_lcia.py`
* `importers/excel.py:`
* `importers/json_ld.py`
* `extractors/csv.py`
* `extractors/excel.py`
* `extractors/json_ld.py`
* `extractors/simapro_csv.py`
* `extractors/simapro_lcia_csv.py`
* `strategies/biosphere.py`
* `strategies/csv.py`
* `strategies/generic.py`
* `strategies/json_ld.py`
* `strategies/json_ld_allocation.py`
* `strategies/json_ld_lcia.py`
* `strategies/simapro.py`

I would like to avoid an unexpected large bill, can we set a weekly cap? Maybe 40 hours per week, this would be OK.
